vendor/spawn/src/spawn_stubs.c

```diff
@@ -480,18 +511,201 @@ CAMLprim value spawn_unix(value v_env,
       switch (sigprocmask_command) {
         case CAML_SIG_SETMASK:
         case CAML_SIG_BLOCK:
-          sigaddset(&info.child_sigmask, signal);
+          sigaddset(&info->child_sigmask, signal);
           break;
 
         case CAML_SIG_UNBLOCK:
-          sigdelset(&info.child_sigmask, signal);
+          sigdelset(&info->child_sigmask, signal);
           break;
 
         default:
           assert(0);
       }
     }
   }
+}
+
+#if defined(USE_POSIX_SPAWN)
+
+CAMLprim value spawn_unix(value v_env,
+                          value v_cwd,
+                          value v_prog,
+                          value v_argv,
+                          value v_stdin,
+                          value v_stdout,
+                          value v_stderr,
+                          value v_use_vfork,
+                          value v_setpgid,
+                          value v_sigprocmask)
+{
+  CAMLparam4(v_env, v_cwd, v_prog, v_argv);
+  CAMLlocal1(e_arg);
+  e_arg = Nothing;
+
+  pid_t pid;
+  int tmp_fds[3] = {0};  // invariant: initialized > 2
+
+  int e_error;
+  char *e_function = NULL;
+
+  posix_spawn_file_actions_t actions;
+  if (posix_spawn_file_actions_init(&actions)) {
+    e_function = "posix_spawn_file_actions_init";
+    goto cleanup;
+  }
+
+  posix_spawnattr_t attr;
+  if (posix_spawnattr_init(&attr)) {
+    e_function = "posix_spawnattr_init";
+    goto cleanup;
+  }
+
+  struct spawn_info info;
+  init_spawn_info(&info, v_env, v_cwd, v_prog, v_argv,
+                  v_stdin, v_stdout, v_stderr, v_setpgid, v_sigprocmask);
+
+  short attr_flags = POSIX_SPAWN_SETSIGMASK;
+  if (info.set_pgid) attr_flags |= POSIX_SPAWN_SETPGROUP;
+  e_error = posix_spawnattr_setflags(&attr, attr_flags);
+  if (e_error) {
+    e_function = "posix_spawnattr_setflags";
+    goto cleanup;
+  }
+
+  e_error = posix_spawnattr_setsigmask(&attr, &info.child_sigmask);
+  if (e_error) {
+    e_function = "posix_spawnattr_setsigmask";
+    goto cleanup;
+  }
+
+  if (info.set_pgid) {
+    e_error = posix_spawnattr_setpgroup(&attr, info.pgid);
+    if (e_error) {
+      e_function = "posix_spawnattr_setpgroup";
+      goto cleanup;
+    }
+  }
+
+  for (int fd = 0; fd < 3; fd++) {
+    int tmp_fd = tmp_fds[fd] = safe_dup(info.std_fds[fd]);
```

Comment:
It's probably worth it to copy the upstream concern about resetting close-on-exec, and think about whether we'd want to solve it here:

https://github.com/janestreet/spawn/pull/47/files#r930889893

Reply:
Yeah, we'll do it upstream. Note that I've talked to the maintainers and we agreed that this isn't an issue for dune because we don't allow anywhere else in the code to use fork.